README.md
<h1 align="center">🗺 map2map-converter 🗺</h1>

<div align="center">

Converter of links to geolocation into coordinates and vice versa without using API services.

[![npm package](https://img.shields.io/npm/v/map2map-converter?logo=npm&style=flat-square)](https://www.npmjs.org/package/map2map-converter)
[![npm download](https://img.shields.io/npm/dm/map2map-converter)](https://www.npmjs.org/package/map2map-converter)
[![https://t.me/vds_13](https://img.shields.io/badge/💬%20Telegram-VDS13-blue.svg?style=flat-square)](https://t.me/vds_13)

</div>

## 📙 Description

The library will help you get the coordinates from the URL without using the API services.

Supported web mapping platforms:
* Yandex
* Google
* 2Gis
* Apple

Supported Telegram bot libraries:
* [Telegraf](https://github.com/telegraf/telegraf)
* [node-telegram-bot-api](https://github.com/yagop/node-telegram-bot-api)
* [telebot](https://github.com/mullwar/telebot)

## 📦 Install

```sh
npm i map2map-converter
```
## [🎚️ Changelog](https://github.com/VDS13/map2map-converter/blob/main/CHANGELOG.md)

## [🗺 API](https://github.com/VDS13/map2map-converter/blob/main/API.md)

## 🚀 Usage

### Example 1:
```js
    const map2map = require('map2map-converter');
    maps = new map2map();
    maps.twoGisMapsToCoord('https://go.2gis.com/7muvw').then((response) => console.log('twoGisMapsToCoord: ' + response.lat + ' ' + response.long));
    maps.yandexMapsToCoord('https://yandex.ru/maps/org/187589077176').then((response) => console.log('yandexMapsToCoord: ' + response.lat + ' ' + response.long));
    maps.googleMapsToCoord('https://maps.app.goo.gl/ijGhmJi4HtQAhE5h8?g_st=it').then((response) => console.log('googleMapsToCoord: ' + response.lat + ' ' + response.long));
    const apple = maps.appleMapsToCoord('https://maps.apple.com/?address=London,%20SW1A%200AA,%20England&auid=8156900488245778404&ll=51.500687,-0.124561&lsp=9902&q=Big%20Ben&_ext=Cj4KBQgEEOEBCgQIBRADCgQIBhBzCgQIChAOCgQIEBABCgQILxBmCgQIUhAFCgQIVRAOCgQIWRAGCgUIpAEQARIkKcCy0qQUwElAMc0uzs5Y6r+/Od7U9mUfwElAQS+175sM2L+/&t=m')
    console.log('coordToYandexMaps: ' + maps.coordToYandexMaps({lat: 50.321, long: 50.123}));
    console.log('coordToGoogleMaps: ' + maps.coordToGoogleMaps({lat: 50.321, long: 50.123}));
    console.log('coordToAppleMaps: ' + maps.coordToAppleMaps({lat: 50.321, long: 50.123}));
    console.log('coordToTwoGisMaps: ' + maps.coordToTwoGisMaps({lat: 50.321, long: 50.123}));
    console.log('identifyMap(google): ' + maps.identifyMap('https://maps.app.goo.gl/sbxrbSUjsfcgU7CG8?g_st=it'));
    console.log('identifyMap(yandex): ' + maps.identifyMap('https://yandex.ru/maps/-/CCUri0hzlC'));
    console.log('identifyMap(2gis): ' + maps.identifyMap('https://go.2gis.com/7muvw'));
    console.log('identifyMap(apple): ' + maps.identifyMap('https://maps.apple.com/?address=London,%20SW1A%200AA,%20England&auid=8156900488245778404&ll=51.500687,-0.124561&lsp=9902&q=Big%20Ben&_ext=Cj4KBQgEEOEBCgQIBRADCgQIBhBzCgQIChAOCgQIEBABCgQILxBmCgQIUhAFCgQIVRAOCgQIWRAGCgUIpAEQARIkKcCy0qQUwElAMc0uzs5Y6r+/Od7U9mUfwElAQS+175sM2L+/&t=m'));
    console.log('appleMapsToCoord: ' + apple.lat + ' ' + apple.long);
```
### Result 1:

<img src="https://github.com/VDS13/map2map-converter/blob/main/doc/example1.jpg"/>

### Example 2:
```js
    const {Telegraf} = require('telegraf');
    const map2map = require('map2map-converter');
    const TOKEN = 'token';

    const bot = new Telegraf(TOKEN);
    const maps = new map2map();

    bot.start((ctx) => maps.coordToTelegramLocation({lat: 50.321, long: 50.123}, ctx.message.chat.id, bot, 'telegraf'));

    bot.on("location", (ctx) => {
        bot.telegram.sendMessage(ctx.message.chat.id ,maps.coordToYandexMaps(maps.telegramLocationToCoord(ctx.message)));
    });
    bot.launch();
    process.once('SIGINT', () => bot.stop('SIGINT'));
    process.once('SIGTERM', () => bot.stop('SIGTERM'));
```
### Result 2:

<img src="https://github.com/VDS13/map2map-converter/blob/main/doc/example2.jpg"/>

## License

**The MIT License (MIT)**

Copyright © 2023 Dmitry Vyazin

CHANGELOG.md

# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

src/map2map.js
const axios = require('axios');
const cheerio = require('cheerio');

module.exports = class map2map {
    identifyMap(url) {
        if ((url.indexOf('goo.gl') !== -1 || url.indexOf('google') !== -1) && url.indexOf('map') !== -1) {
            return 'google';
        } else if (url.indexOf('yandex') !== -1 && url.indexOf('map') !== -1) {
            return 'yandex';
        } else if (url.indexOf('apple') !== -1 && url.indexOf('map') !== -1) {
            return 'apple';
        } else if (url.indexOf('2gis') !== -1) {
            return '2gis';
        }
        return -1;
    }
    async yandexMapsToCoord(url) {
        let response = await axios({
            method: 'GET',
            url: url,
            headers:{
                "Accept": "application/json, text/javascript, */*; q=0.01",
                "Accept-Language": "ru",
                "Cache-Control": "no-cache",
                "Connection": "keep-alive",
                "Cookie": '',
                "DNT": "1",
                "Host": "yandex.ru",
                "Pragma": "no-cache",
                "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.102 Safari/537.36",
                "X-Requested-With": "XMLHttpRequest",
            }
        });
        if (url.indexOf('maps/-/') !== -1) {
            const $ = cheerio.load(response.data);
            const res = $("div.toponym-card-title-view__coords-badge").text().split(', ');
            if (res[1] !== undefined)
                return {lat: res[0], long: res[1]};
        }  else if (url.indexOf('maps/org/') !== -1) {
            const res = response.data.match(/displayCoordinates\":\[(.(-?)[0-9]{1,15})(.(-?)[0-9]{1,15})+/)[0].substr(21).split(',');
            if (res[1] !== undefined)
                return {lat: res[1], long: res[0]};
        }
        return {lat: 'err', long: 'err'};
    }
    async googleMapsToCoord(url) {
        let response = await axios.get(url);
        const res = response.data.match(/APP_INITIALIZATION_STATE=\[\[\[((-?)[0-9][0-9]*)(.(-?)[0-9]{1,15})+/)[0].substr(28).split(',');
        if (res[1] !== undefined)
            return {lat: res[2], long: res[1]};
        return {lat: 'err', long: 'err'};
    }
    async twoGisMapsToCoord(url) {
        let response = await axios.get(url);
        const $ = cheerio.load(response.data);
        const res = new URL($("input[name=return_url]").val()).pathname.split('/').slice(-1)[0].split('%2C');
        return {lat: res[1], long: res[0]};
    }
    appleMapsToCoord(url) {
        const res = new URL(url).searchParams.get('ll').split(',');
        return {lat: res[0], long: res[1]};
    }
    telegramLocationToCoord(msg) {
        return {lat: msg.location.latitude, long: msg.location.longitude};
    }
    coordToYandexMaps(coord) {
        return 'https://yandex.ru/maps/?ll=' + coord.long + '%2C' + coord.lat + '&mode=whatshere&whatshere%5Bpoint%5D=' + coord.long + '%2C' + coord.lat + '&whatshere%5Bzoom%5D=18&z=18';
    }
    coordToTwoGisMaps(coord) {
        return 'https://2gis.ru/geo/' + coord.long + '%2C' + coord.lat;
    }
    coordToGoogleMaps(coord) {
        return 'https://maps.google.com/?q=' + coord.lat + ',' + coord.long;
    }
    coordToAppleMaps(coord) {
        return 'https://maps.apple.com/place?ll=' + coord.lat + '%2C' + coord.long;
    }
    coordToTelegramLocation(coord, chat_id, bot, bot_type) {
        if (bot_type === 'node-telegram-bot-api') {
            bot.sendLocation(chat_id, coord.lat, coord.long);
        } else if (bot_type === 'telegraf') {
            bot.telegram.sendLocation(chat_id, coord.lat, coord.long);
        } else if (bot_type === 'telebot') {
            bot.sendLocation(chat_id, [coord.lat, coord.long]);
        }
    }
}

example/example.js
const http = require('http');
const map2map = require('map2map-converter');
const hostname = '127.0.0.1';
const port = 3000;
const axios = require('axios');

const server = http.createServer((req, res) => {
    res.statusCode = 200;
    res.setHeader('Content-Type', 'text/plain');
    res.end('Hello World');
    
});

server.listen(port, hostname, () => {
    console.log(`Server running at http://${hostname}:${port}/`);
    maps = new map2map();
    maps.twoGisMapsToCoord('https://go.2gis.com/7muvw').then((response) => console.log('twoGisMapsToCoord: ' + response.lat + ' ' + response.long));
    maps.yandexMapsToCoord('https://yandex.ru/maps/org/187589077176').then((response) => console.log('yandexMapsToCoord: ' + response.lat + ' ' + response.long));
    maps.googleMapsToCoord('https://maps.app.goo.gl/ijGhmJi4HtQAhE5h8?g_st=it').then((response) => console.log('googleMapsToCoord: ' + response.lat + ' ' + response.long));
    const apple = maps.appleMapsToCoord('https://maps.apple.com/?address=London,%20SW1A%200AA,%20England&auid=8156900488245778404&ll=51.500687,-0.124561&lsp=9902&q=Big%20Ben&_ext=Cj4KBQgEEOEBCgQIBRADCgQIBhBzCgQIChAOCgQIEBABCgQILxBmCgQIUhAFCgQIVRAOCgQIWRAGCgUIpAEQARIkKcCy0qQUwElAMc0uzs5Y6r+/Od7U9mUfwElAQS+175sM2L+/&t=m')
    console.log('coordToYandexMaps: ' + maps.coordToYandexMaps({lat: 50.321, long: 50.123}));
    console.log('coordToGoogleMaps: ' + maps.coordToGoogleMaps({lat: 50.321, long: 50.123}));
    console.log('coordToAppleMaps: ' + maps.coordToAppleMaps({lat: 50.321, long: 50.123}));
    console.log('coordToTwoGisMaps: ' + maps.coordToTwoGisMaps({lat: 50.321, long: 50.123}));
    console.log('identifyMap(google): ' + maps.identifyMap('https://maps.app.goo.gl/sbxrbSUjsfcgU7CG8?g_st=it'));
    console.log('identifyMap(yandex): ' + maps.identifyMap('https://yandex.ru/maps/-/CCUri0hzlC'));
    console.log('identifyMap(2gis): ' + maps.identifyMap('https://go.2gis.com/7muvw'));
    console.log('identifyMap(apple): ' + maps.identifyMap('https://maps.apple.com/?address=London,%20SW1A%200AA,%20England&auid=8156900488245778404&ll=51.500687,-0.124561&lsp=9902&q=Big%20Ben&_ext=Cj4KBQgEEOEBCgQIBRADCgQIBhBzCgQIChAOCgQIEBABCgQILxBmCgQIUhAFCgQIVRAOCgQIWRAGCgUIpAEQARIkKcCy0qQUwElAMc0uzs5Y6r+/Od7U9mUfwElAQS+175sM2L+/&t=m'));
    console.log('appleMapsToCoord: ' + apple.lat + ' ' + apple.long);
});
